ra-optimisation/search-tree-rao/src/main/java/com/powsybl/openrao/searchtreerao/searchtree/parameters/SearchTreeParameters.java

```diff
@@ -134,6 +137,63 @@ public void setRaLimitationsForSecondPreventive(RaUsageLimits raUsageLimits, Set
+                        decreasedRaUsageLimits.setMaxTso(decreasedMaxTso);
+
+                        raLimitationParameters.put(otherInstant, decreasedRaUsageLimits);
+                    }
+                }
+            )
+        );
+    }
+
+    private static int decreaseMaxRemedialAction(RaUsageLimits raUsageLimits, State optimizedState, OptimizationResult result) {
+        return raUsageLimits.getMaxRa() - result.getActivatedNetworkActions().size() - result.getActivatedRangeActions(optimizedState).size();
+    }
+
+    private static Map<String, Integer> decreaseMaxTopoPerTso(RaUsageLimits raUsageLimits, OptimizationResult result, Map<String, Integer> decreasedMaxRaPerTso) {
+        Map<String, Integer> decreasedMaxTopoPerTso = new HashMap<>();
+        raUsageLimits.getMaxTopoPerTso().forEach((key, value) -> decreasedMaxTopoPerTso.put(key, Math.min(value - result.getActivatedNetworkActions().stream().filter(networkAction -> key.equals(networkAction.getOperator())).toList().size(), decreasedMaxRaPerTso.get(key))));
+        return decreasedMaxTopoPerTso;
+    }
+
+    private static Map<String, Integer> decreaseMaxPstPerTso(RaUsageLimits raUsageLimits, State optimizedState, OptimizationResult result, Map<String, Integer> decreasedMaxRaPerTso) {
+        Map<String, Integer> decreasedMaxPstPerTso = new HashMap<>();
+        raUsageLimits.getMaxPstPerTso().forEach((key, value) -> decreasedMaxPstPerTso.put(key, Math.min(value - result.getActivatedRangeActions(optimizedState).stream().filter(networkAction -> key.equals(networkAction.getOperator())).toList().size(), decreasedMaxRaPerTso.get(key))));
+        return decreasedMaxPstPerTso;
+    }
+
+    private static Map<String, Integer> decreaseMaxRemedialActionPerTso(RaUsageLimits raUsageLimits, State optimizedState, OptimizationResult result) {
+        Map<String, Integer> decreasedMaxRaPerTso = new HashMap<>();
+        raUsageLimits.getMaxRaPerTso().forEach((key, value) -> decreasedMaxRaPerTso.put(key, value - result.getActivatedNetworkActions().stream().filter(networkAction -> key.equals(networkAction.getOperator())).toList().size() - result.getActivatedRangeActions(optimizedState).stream().filter(networkAction -> key.equals(networkAction.getOperator())).toList().size()));
+        return decreasedMaxRaPerTso;
+    }
+
+    private static int decreaseMaxTso(RaUsageLimits raUsageLimits, State optimizedState, OptimizationResult result) {
+        int tsosWithPreviouslyAppliedRas = raUsageLimits.getMaxTsoExclusion().size();
+        result.getActivatedNetworkActions().forEach(networkAction -> raUsageLimits.addTsoToExclude(networkAction.getOperator()));
+        result.getActivatedRangeActions(optimizedState).forEach(rangeAction -> raUsageLimits.addTsoToExclude(rangeAction.getOperator()));
+        return raUsageLimits.getMaxTso() - (raUsageLimits.getMaxTsoExclusion().size() - tsosWithPreviouslyAppliedRas);
```

Comment:
I don't understand this line of code

Reply:
The maximum number of TSOs is decreased only by the number of new TSOs with RAs that were applied at the last instant)